gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -79,59 +88,122 @@ public LoopingDatasetFinderSource(boolean drilldownIntoPartitions) {
     return Lists.newArrayList(getWorkunitStream(state).getMaterializedWorkUnitCollection());
   }
 
+  /**
+   * Collect all failed workunit in previous run into a set.
+   *
+   * Workunits that exceed max retry limit will not be added into failedPreviousWorkUnits
+   * and rollovered into next quota-available execution.
+   *
+   * @param retriedUrns Track all dataset/partition's urns that is included in the retried workunits, so that in the
+   *                    case where the number of remaining dataset/partitions are less than the maximum number of
+   *                    workunits can be processed by single execution, a retried workunit won't be processed twice
+   */
+  private List<WorkUnit> getPreviousFailedWorkUnits(List<WorkUnitState> previousWorkUnitStates, int maxRetries,
+      Set<String> retriedUrns) {
+    List<WorkUnit> failedPreviousWorkUnits = new ArrayList<>();
+    for (WorkUnitState workUnitState : previousWorkUnitStates) {
+      if ((workUnitState.getWorkingState() == WorkUnitState.WorkingState.FAILED
+          || workUnitState.getWorkingState() == WorkUnitState.WorkingState.ROLLOVER)) {
+        if (!workUnitState.contains(ConfigurationKeys.TASK_RETRIES_KEY)
+            || workUnitState.getPropAsInt(ConfigurationKeys.TASK_RETRIES_KEY) < maxRetries + 1) {
+          log.info("Dataset " + workUnitState.getProp(DATASET_URN) + " is failed previously, retrying...");
+          WorkUnit retryWorkUnit = new WorkUnit(workUnitState.getWorkunit());
+          failedPreviousWorkUnits.add(retryWorkUnit);
+          String retriedUrn = workUnitState.getWorkunit().contains(PARTITION_URN) ? retryWorkUnit.getProp(PARTITION_URN)
+              : retryWorkUnit.getProp(DATASET_URN);
+          log.info("The workunit: " + retriedUrn + " is added into retry bucket.");
+          retriedUrns.add(retriedUrn);
+        } else {
+          log.warn("Dataset " + workUnitState.getProp(DATASET_URN) + " exceeds retry limit : " + maxRetries);
+        }
+      }
+    }
+    return failedPreviousWorkUnits;
+  }
+
+  /**
+   * Searching for workunit with maximum ordinal.
+   * @param previousWorkUnitStates
+   * @return
+   */
+  private Optional<WorkUnitState> getMaxWorkUnitState(List<WorkUnitState> previousWorkUnitStates) {
+    int tempWorkUnitOrdinal = 0;
+    Optional<WorkUnitState> optionalMaxWorkUnit = Optional.empty();
+    for (WorkUnitState workUnitState : previousWorkUnitStates) {
+      try {
+        if (workUnitState.getPropAsInt(WORK_UNIT_ORDINAL) > tempWorkUnitOrdinal) {
+          optionalMaxWorkUnit = Optional.of(workUnitState);
+          tempWorkUnitOrdinal = workUnitState.getPropAsInt(WORK_UNIT_ORDINAL);
+        }
+      } catch (NumberFormatException nfe) {
+        throw new RuntimeException(
+            "Work units in state store are corrupted! Missing or malformed " + WORK_UNIT_ORDINAL);
```

Comment:
Please include the `nfe` in the `RuntimeException`. Whenever possible, please always include nested exceptions.

`throw new RuntimeException("...message...", nfe)`